ext/pdo_pgsql/pgsql_driver.c

```diff
@@ -1224,6 +1242,34 @@ PHP_METHOD(PDO_PGSql_Ext, pgsqlGetPid)
 }
 /* }}} */
 
+/* {{{ proto bool PDO::pgsqlSetNoticeCallback(mixed callback)
+   Sets a callback to receive DB notices (after client_min_messages has been set) */
+PHP_METHOD(PDO_PGSql_Ext, pgsqlSetNoticeCallback)
+{
+	pdo_dbh_t *dbh;
+	pdo_pgsql_db_handle *H;
+	zend_fcall_info fci = empty_fcall_info;
+	zend_fcall_info_cache fcc = empty_fcall_info_cache;
+	
+	dbh = Z_PDO_DBH_P(getThis());
+	PDO_CONSTRUCT_CHECK;
+	
+	H = (pdo_pgsql_db_handle *)dbh->driver_data;
+	
+	pdo_pgsql_cleanup_notice_callback(H);
```

Comment:
Cleaning up the notice callback should only be done _after_ the parameter parsing succeeded.

Reply:
Hmm, I thought that, would an error occur, it would have at least done half the work and let the object without callback; that is, a little bit more clean than not even detaching the previous callback. Which isn't a great deal, granted, because at this point (which probably ends up in a fatal) "letting a clean place" is not our main concern.

But there's another benefit: if our hungry FPM process is in great danger of crashing, having consumed all but a handful of bytes over its allowed 0x20000000, and tries to cleanup _after_ parsing, it may fail the parsing.
On the other hand comes our "cleanup _before_ parsing" saver, that frees just enough bytes for parsing to then succeed, allow the process to reach a mem release checkpoint and then cross the finish line (but with no photographer at all witnessing its performance, because they all would be affaired around the process that embarassingly crashed just inches away).

… I won't provide the test case for this hypothesis 😄 but you get the idea.

Reply:
I don't think it would benefit the memory limit in most cases because memory is allocated in chunk anyway.
Validating arguments should also not influence memory usage, unless a type error occurs but even then...
We normally always place the actual work after parsing, so that should be done here too :slightly_smiling_face:

> I just found 1 case of PDO_CONSTRUCT_CHECK before ZPP, in pdo_sqlite.c's openBlob

I think that should be fixed.